app/main.py

```diff
@@ -1,3 +1,44 @@
-def cache(func):
-    # Write your code here
-    pass
+from functools import wraps
+from typing import Callable, Any
+
+
+def cache(func: Callable) -> Callable:
+    caching_result = {}
+    wraps(func)
+
+    def wrapper(*args: tuple) -> Any:
+        nonlocal caching_result
+
+        if func.__name__ not in caching_result.keys():
+            print("Calculating new result")
+
+            result = func(*args)
+
+            caching_result[func.__name__] = [{
+                "arguments": args,
+                "result": result
+            }]
+
+            return result
+        elif args not in [
+            value["arguments"]
+            for value in caching_result[func.__name__]
+        ]:
+            print("Calculating new result")
+
+            result = func(*args)
+
+            caching_result[func.__name__].append({
+                "arguments": args,
+                "result": result
+            })
+
+            return result
+        else:
+            print("Getting from cache")
+
+            for item in caching_result[func.__name__]:
+                if item["arguments"] == args:
+                    return item["result"]
```

Comment:
Your logic here is overloaded. You do not need to store the function name as a key for `caching_result`, you could use `args` directly as a key and the function result as a value. This cache decorator will be applied to one function at a time, it doesn't need to track all functions it was applied to. After that, you can check if the same args are already present in the `caching_result` and either retrieve it from there or put a new value. Besides that, try to use only one `return` statement in the `wrapper` and do not call `nonlocal caching_result`, it will be available in the `wrapper` context.